src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp

```diff
@@ -86,22 +76,114 @@ class SSELoss
   template<bool UseWeights, typename MatType, typename WeightVecType>
   double Evaluate(const MatType& input, const WeightVecType& /* weights */)
   {
+    // Sanity check for empty input.
+    if (input.n_cols == 0)
+      return 0.0;
+
     // Calculate gradients and hessians.
     gradients = (input.row(1) - input.row(0)).t();
     hessians = arma::vec(input.n_cols, arma::fill::ones);
 
-    return std::pow(ApplyL1(arma::accu(gradients)), 2) /
-        (arma::accu(hessians) + lambda);
+    // Calculates total sum of gradients and hessians.
+    gTotal = arma::accu(gradients);
+    hTotal = arma::accu(hessians);
+
+    return std::pow(ApplyL1(gTotal), 2) / (hTotal + lambda);
+  }
+
+  /**
+   * Sorts the gradients and hessians according to the sorted order of the
+   * feature.
+   *
+   * @param sortedIndices Vector of indices according to the sorted order of
+   *      the feature.
+   * @param sortedGradients Vector to store the sorted gradients.
+   * @param sortedHessians Vector to store the sorted hessians.
+   */
+  void sortGradAndHess(const arma::uvec& sortedIndices,
+                       arma::vec& sortedGradients,
+                       arma::vec& sortedHessians)
+  {
+    for (size_t i = 0; i < sortedIndices.n_elem; ++i)
+    {
+      sortedGradients[i] = gradients[sortedIndices[i]];
+      sortedHessians[i] = hessians[sortedIndices[i]];
+    }
+  }
+
+  /**
+   * Calculates the sum of hessians and gradients for the left child.
+   *
+   * @return The start index for the iteration to find the best split.
+   */
+  size_t BinaryScanInitialize(const arma::vec& sortedGradients,
+                              const arma::vec& sortedHessians)
+  {
+    size_t index = 0;
+    // Initializing the data members.
+    gLeft = 0;
+    hLeft = 0;
+    while (index < sortedHessians.n_elem - 1 &&
+           hLeft + sortedHessians[index] < minChildWeight)
+    {
+      gLeft += sortedGradients[index];
+      hLeft += sortedHessians[index++];
+    }
+    return index;
+  }
+
+  /**
+   * Steps through the current index and update the statistics. While taking
+   * the step, it ensures that the weight of right child is higher than the
+   * minChildWeight.
+   *
+   * @return Returns true if the current step is valid else returns false.
+   */
+  bool BinaryStep(const arma::vec& sortedGradients,
+                  const arma::vec& sortedHessians,
+                  size_t& index)
+  {
+    gLeft += sortedGradients[index];
+    hLeft += sortedHessians[index++];
+
+    // If the total weight of right child is lesser than the minChildWeight,
+    // then we need to terminate the loop for iterating over split points.
+    if (hTotal - hLeft < minChildWeight)
+      return false;
+    return true;
+  }
+
+  /**
+   * Calculates the gains for the current split.
+   */
+  double BinaryGains()
```

Comment:
Shouldn't this return `std::tuple<double, double>`?

Reply:
If we stick to the interface we had in `BestBinaryNumericSplit`, then yes, it should return a `std::tuple`. As you saw that there are minor changes in the signatures of `BinaryScanInitialize` and `BinaryStep` too. I did this so that the code here looks a bit cleaner.

> I think that change is fine, but can you also update the other loss functions and BestBinaryNumericSplit accordingly?

I am wondering why do we actually need to make both of them look exactly the same? The reason I did those checks in those functions was because I can't access `minChildWeight` from outside. If that were the case, then I would have definitely done that within the splitting strategy class. If we try to make these similar, then we have to do these checks inside the functions, but personally, I think it makes more sense to have them within the loop itself because the checks are a part of the splitting strategy and not the LossFunction or Fitnessfunction. I had to do that inside LossFunction here because I had no other choice than that.

Let me know if I understood you correctly. (I am having a feeling that you are trying to say something else and I have replied something else) :)

Reply:
The purpose of an abstraction like we are using here is to logically separate two concepts by means of a pre-defined interface.  That interface, in this case, is the functions `BinaryScanInitialize()`, `BinaryStep()`, and `BinaryGains()` (and their signatures).  If we have a collection of gain functions that implement *similar* but not *identical* interfaces, then we cannot plug and play different losses.

Things are a little bit different because we are using C++ templates, but for the sake of discussion, imagine we were using inheritance: we would define a base `Gain` class, and then overload the various methods in the children.  This way, the thing that used the gain (i.e. `XGBExactNumericSplit`, `BestBinaryNumericSplit`, etc.) could just call the functions defined in the base `Gain` class, regardless of the true derived type of the gain being used.

We should work very hard to make clean abstractions, as that makes later extension easy.  It also makes testing easier, and helps make the code easier to approach and understand because there are fewer linkages between sections of code.

I want it to be possible to use `SSELoss` with `BestBinaryNumericSplit`; and similar, I want it to be possible to use `MSEGain` with `XGBExactNumericSplit`.  If we need to change the signatures of the functions involved with the abstraction, that's okay, but we need to make sure things are sufficiently general.

In the case of `minChildWeight`, this seems very specific to `SSELoss` and I think I have to agree that we can only realistically check that inside of `SSELoss` itself.  Otherwise we would have code specific to `SSELoss` inside of `XGBExactNumericSplit`.

Reply:
Let me know if there are any "problematic" linkages between `SSELoss` and `XGBExactNumericSplit` that seem hard to overcome.  It should definitely be possible, but sometimes it can be hard to figure out exactly how to separate things cleanly. :)

Reply:
Thanks for the comprehensive write-up. I too agree that we need to make the interfaces as general as possible. But, I have one question.

> I want it to be possible to use SSELoss with BestBinaryNumericSplit; and similar, I want it to be possible to use MSEGain with XGBExactNumericSplit

Why do we need to make this possible in the first place? I mean `SSELoss` has a regularized objective, have some hyperparameters which are specific only to xgboost and calculates gain & output values using second-order taylor expansion. One can never use standard gain computation classes that are built for `BestBinaryNumericSplit` in `XGBExactNumericSplit` because it won't meet it's requirements. Now if we talk the other way round i.e. using `SSELoss` in `BestBinaryNumericSplit`, we actually don't have the appropriate data members stored within the `SSELoss` class thus we would need to add some additional data members which might not always be needed.

The point I am trying to make is that `SSELoss` is fundamentally different in the way it works. Though on the higher level it may seem that at the end of the day it is just calculating gain and output values just like any other gain function we have for decision trees, but the internal workings are very different. I would even say that it would be logically wrong to use the `SSELoss` function anywhere other than any "tree boosting" method.

> Let me know if there are any "problematic" linkages between SSELoss and XGBExactNumericSplit that seem hard to overcome.

I agree with that. I can try to decrease the interdependence of `XGBExactNumericSplit` and `SSELoss` so that we can better follow the policy-based design. I will try to look for some viable options for this :)

Reply:
>> I want it to be possible to use SSELoss with BestBinaryNumericSplit; and similar, I want it to be possible to use MSEGain with XGBExactNumericSplit
>
> Why do we need to make this possible in the first place? 

For the sake of generality---from a conceptual level, there is actually no reason why I couldn't use `SSELoss` with `BestBinaryNumericSplit` or any other splitting strategy.  The splitting strategy is independent of the loss function used, regardless of whether we are talking about XGBoost or not.  In fact, when I look through the `XGBExactNumericSplit` class, I think that it is actually unnecessary: all `XGBExactNumericSplit` is doing (as far as I can tell but it's possible I missed a detail) is finding the best binary split.  So, if all of the details about gradients and hessians (which are specific to the loss function, not the splitting strategy) are properly encapsulated inside of `SSELoss`, then we can just use `BestBinaryNumericSplit` directly and it will do the same thing.